pkg/services/provisioning/dashboards/file_reader_test.go

```diff
@@ -101,55 +100,97 @@ func TestDashboardFileReader(t *testing.T) {
 				reader, err := NewDashboardFileReader(cfg, logger)
 				So(err, ShouldBeNil)
 
+				bus.AddHandler("test", func(cmd *models.GetDashboardQuery) error {
+					return models.ErrDashboardNotFound
+				})
+
+				fakeService.
+					On(
+						"SaveFolderForProvisionedDashboards",
```

Comment:
I'm sorry but for me personally I'm not really seeing why this is better than the old way. No offense, but seems to me like more code and some magic strings are used and you don't get anything extra that we really need? 

Really good initiative though for get the discussions started :+1:

Reply:
Yup I agree that it looks worse, Go does not seem to provide a great way to mock stuff. But the main point is how you change and extend this. For example when I had to add `DeleteDashboard` method I had to:

- implement it to modify the internal state of the mock, which in this case was simple, in other cases could be more complicated and could introduce error.
- After that you need to check if the internal state is correct in the test after the delete (ideally diff it with expected state, I don't think I did that as these tests just check number of items in the internal state not if correct item was deleted)

With mocking like testify you do:
- add the method into the mock in the same way as others so I think less space for error
- add something like 
```
fake.On("Delete", 1).Return(nil) // test that it is trying to delete thing with id 1
```

If you want to add error handling for test the changes would be:
Manual mock:
- Add some switch to control what the mock returns when called and setup it in the test. I seen different ways this is implemented so there does not seem some common API for this but I assume something like `fake.returnOnGetDashboards = error` and check for that in the function?

Testify:
- `fake.On("GetDashboards", mock.Anything).Return(error)`

So all in all this gives you some common API with documentation for common mocking usecases (check what was called with what arguments, inject what should be returned).

The manual mocks give you some functionality for free, like if you insert Dashboard with `SaveDashboard` next call to `GetDashboards` will retrieve them without you needing to setup that. I see the appeal in that but again you have to be sure this will hold true and there is no error in the mock which there can be if complexity goes up. If you need to add some switch for returning errors there will be additional API that can be inconsistent with other mocks. Also the API for injecting return value into `GetDashboards` call is there, it is either `SaveDashboard` or internal `mock.dashboards` state that you need to manipulate.

So even thought these aren't blocking issues, they sway me in the direction of one mocking library with unified and documented API even for the price of more verbose code. I assume there are usecases for which manual mocks with internal state will be more useful and this does not prevent using them but I think in most usecases you will be better of with simple `on calling this return this` kind of functionality.

Reply:
A specific case of this I need to solve right now in #16211. I have a function in the fake implementation of `dashboardService`:
```go
func (s *FakeDashboardService) DeleteDashboard(dashboardId int64, orgId int64) error {
	for index, dash := range s.SavedDashboards {
		if dash.Dashboard.Id == dashboardId && dash.OrgId == orgId {
			s.SavedDashboards = append(s.SavedDashboards[:index], s.SavedDashboards[index+1:]...)
			break
		}
	}
	return nil
}
```
But I added a check to real version so that if dashboard is provisioned it cannot be deleted. If I want to test this from dashboard API tests where this mock is used, I need to implement a way to inject the error here, either by also implementing the check here and also changing the internal state to contain the provisioned flag in some way so it can be set or I need to add another flag and do it in a way like here:
```go
func (s *FakeDashboardService) SaveDashboard(dto *SaveDashboardDTO) (*models.Dashboard, error) {
	s.SavedDashboards = append(s.SavedDashboards, dto)

	if s.SaveDashboardResult == nil && s.SaveDashboardError == nil {
		s.SaveDashboardResult = dto.Dashboard
	}

	return s.SaveDashboardResult, s.SaveDashboardError
}
```
where `SaveDashboardError` stores the error to be returned. BTW not sure if there isn't a bug in there as this line `s.SavedDashboards = append(s.SavedDashboards, dto)` is executed even if `SaveDashboardError` is set actually saving the data in internal state even though it errors out.

Reply:
> But I added a check to real version so that if dashboard is provisioned it cannot be deleted. If I want to test this from dashboard API tests where this mock is used, I need to implement a way to inject the error here

To me that sounds that you shouldn't test using a mock of the `DashboardProvisioningService` - instead the real implementation and add a bus mock handler for the `DashboardProvisioningService.DeleteProvisionedDashboard`

Reply:
Yes that could be a solution in this example. And I would say it makes sense to have more integration tests (maybe even hitting database). On the other hand changing unit test to integration test changes the granularity of the tests which is not always what you need. Also there are instances when you need to mock out something else than DB, like IO operations, time sensitive code or calling external API which would either need to be done over the bus (not sure if we want to) for bus mocking to be universally applicable or we would inevitably need to do some mocks anyway.